src/key_store/mod.rs

```diff
@@ -0,0 +1,184 @@
+//! A storage solution for cryptographic key material.
+//!
+//! This module provides access to the `KeyStore` struct, which manages the
+//! storage of `CredentialBundle` and `KeyPackageBundle` instances. The
+//! development of this module is tracked in #337, which also includes a
+//! roadmap.
+//!
+//! The current key store enables the storage of `CredentialBundle` instances,
+//! and grants access to `CredentialBundle` references via the
+//! `SignaturePublicKey` of the corresponding `Credential`.
+//!
+//! A `KeyStore` is meant to be used across multiple `ManagedGroup` instances to
+//! allow sharing the same `CredentialBundle`. If this is not desired, multiple
+//! `KeyStore` instances can be used across groups.
+//!
+//! # Example
+//!
+//! A simple example for the generation and the retrieval of a
+//! `CredentialBundle`.
+//!
+//! ```
+//! use openmls::prelude::*;
+//!
+//! let key_store = KeyStore::default();
+//!
+//! // Generate a credential bundle.
+//! let alice_credential = key_store
+//!     .generate_credential(
+//!         "Alice".into(),
+//!         CredentialType::Basic,
+//!         SignatureScheme::ED25519,
+//!     )
+//!     .unwrap();
+//!
+//! // Generate a key package bundle with a matching ciphersuite.
+//! let ciphersuite_name = CiphersuiteName::MLS10_128_DHKEMX25519_AES128GCM_SHA256_Ed25519;
+//!
+//! let alice_key_package = key_store
+//!     .generate_key_package(&[ciphersuite_name], &alice_credential, vec![])
+//!     .unwrap();
+//!
+//! // Create a group with the previously generated credential and key package.
+//! let managed_group_config = ManagedGroupConfig::new(
+//!     HandshakeMessageFormat::Plaintext,
+//!     UpdatePolicy::default(),
+//!     0,
+//!     0,
+//!     ManagedGroupCallbacks::default(),
+//! );
+//!
+//! let alice_group = ManagedGroup::new(
+//!     &key_store,
+//!     &managed_group_config,
+//!     GroupId::from_slice(b"Test Group"),
+//!     &alice_key_package.hash(),
+//! )
+//! .unwrap();
+//! ```
+use std::{
+    collections::HashMap,
+    sync::{RwLock, RwLockReadGuard},
+};
+
+use crate::{
+    ciphersuite::{CiphersuiteName, SignaturePublicKey, SignatureScheme},
+    credentials::{Credential, CredentialBundle, CredentialType},
+    extensions::Extension,
+    key_packages::{KeyPackage, KeyPackageBundle},
+};
+
+pub mod errors;
+
+#[cfg(test)]
+mod test_key_store;
+
+pub use errors::KeyStoreError;
+
+#[derive(Debug, Default)]
+/// The `KeyStore` contains `CredentialBundle`s and `KeyPackageBundle`s and
+/// makes them available via information in the corresponding `Credential` and
+/// `KeyPackage` instances.
+pub struct KeyStore {
+    // Map from signature public keys to credential bundles
+    credential_bundles: RwLock<HashMap<SignaturePublicKey, RwLock<CredentialBundle>>>,
```

Comment:
So something that gives you a lot of trouble is the value in the hash map here. Why do you need to have an `RwLock` on it?

Reply:
Excellent question! It used to be a `RefCell`, because at some point, I think I convinced myself that we needed another layer of interior mutability here. But on further inspection, I think it might not actually be needed. I've removed it and it turns out it's much nicer without it 🙂

Thanks for pointing this out!